beacon_node/lighthouse_network/src/discovery/mod.rs

```diff
@@ -1036,12 +1038,107 @@ impl<TSpec: EthSpec> NetworkBehaviour for Discovery<TSpec> {
+
+                trace!(self.log, "Received NewListenAddr event from swarm"; "listener_id" => ?listener_id, "addr" => ?addr);
+
+                let ip: IpAddr = match addr.iter().next() {
+                    Some(Protocol::Ip4(ip)) => IpAddr::V4(ip),
+                    Some(Protocol::Ip6(ip)) => IpAddr::V6(ip),
+                    _ => {
+                        debug!(self.log, "Encountered unacceptable multiaddr for listening (no IP)"; "addr" => ?addr);
+                        return;
+                    }
+                };
+
+                let update_enr = |ip: IpAddr, transport: Protocol| {
+                    /* this closure shouldn't exist but `Ipv4Addr::is_global` is Nightly at the moment */
+                    let is_global = |ip: Ipv4Addr| {
+                        !(ip.is_broadcast()
+                            || ip.is_documentation()
+                            || ip.is_link_local()
+                            || ip.is_loopback()
+                            || ip.is_multicast()
+                            || ip.is_private()
+                            || ip.is_unspecified())
+                    };
+
+                    let insert_into_enr = |key: &str, port: &u16| {
+                        if let Err(e) = self.discv5.enr_insert(key, port) {
+                            warn!(self.log, "Failed to write to ENR"; "error" => ?e);
+                        }
+                    };
+
+                    let mut should_update_ip: bool = true;
+
+                    /* we cannot violate the ENR invariant */
+                    if let IpAddr::V4(ip4) = ip {
+                        if !is_global(ip4) {
+                            debug!(self.log, "Refusing to update ENR IP (unreachable)"; "ip" => ?ip);
+                            should_update_ip = false;
+                        }
+                    }
+
+                    let port: u16 = match transport {
+                        Protocol::Tcp(port) => port,
+                        Protocol::Udp(port) => port,
+                        _ => {
+                            crit!(self.log, "Attempt to update discovery with unsupported transport (this should never occur)"; "transport" => ?transport);
+                            return;
+                        }
+                    };
+
+                    let transport_is_tcp: bool = matches!(transport, Protocol::Tcp(_));
+
+                    if should_update_ip {
+                        let sock: SocketAddr = SocketAddr::new(ip, port);
+
+                        self.discv5.update_local_enr_socket(sock, transport_is_tcp);
+                        *self.network_globals.local_enr.write() = self.discv5.local_enr();
+                    } else {
+                        match ip {
+                            IpAddr::V4(_) => {
+                                if transport_is_tcp {
+                                    insert_into_enr("tcp4", &port);
```

Comment:
We also probably should check that the tcp port isn't already set from a previous run. I.e we load it from disk. I can't recall maybe the insert() function already checks. But we probably dont want to be increasing the sequence number if we don't have to.

Reply:
> We also probably should check that the tcp port isn't already set from a previous run. I.e we load it from disk.

Are you referring to previous instances of us entering this swarm event handler throughout the duration of the BN runtime? Namely, if we receive multiple local addresses to listen on from the swarm (which virtually always happens during normal operation), when we perform the updates serially, we keep thrashing the ENR's sequence number.

Reply:
I was thinking of the case like, we run Lighthouse on port 9000 as default. It creates an ENR from fresh because its our first ever time running. We set the tcp port on the ENR. 

We then shutdown lighthouse and save the ENR to disk. 

Then we run lighthouse again. This loads the saved ENR from disk which has the TCP field already set. If we are using the exact same CLI params (i.e all we have done is restart lighthouse), the tcp field in the ENR is set to the one that libp2p would return to us so we probably don't need to reset it.